rfps/frontrunning_protection_parachain.md

```diff
@@ -0,0 +1,116 @@
+
+We need to de-reference all Commits relevant to a finalised block and apply their extrinsic in order that they were Committed (else we are yet again subject to a frontrunning attack):
+1. Get all Reveal Txs from a finalised block
+2. For each Reveal, find its corresponsing Commit (we might need to look in multiple different blocks, but not too far ahead since Txs are limited by TTL)
+3. Order the Commit Txs by block number (and if multiple within block, by placement in the extrinsic list)
+4. Following the Commit Txs order, apply the extrinsics from Reveal Txs, hence modifying the storage.
+
+
+### Block production mechanism
+
+Under the proposed mechanism, a Commit transaction can be included in any block by the active parachain collators.
+
+All Commit transactions have a short Time-To-Live (TTL) as a mechanism to prevent spamming. The concrete scenario to circumvent is that of collators bloating the blocks by including a range of potentially beneficial Commit Txs, and only selectively publishing Reveals which benefit them.
+
+A Reveal Tx, however, should only be put into a block if there is a corresponding alive Commit Tx that attests the validity of the Reveal Tx. It is required for the Commit Tx to already be in a finalised block to be valid (at time of submission)
+
+A collator is free to include invalid Tx in a block, but these should be rejected by the peers (and relay chain validators) given the invalidity of transactions it contains.
+
+### Refundable Commit fee
+In order to submit a Commit Tx, a user must attach a "deposit" - a high, refundable fee that will be returned upon successful reveal.
+
+Commit Tx contents are secret, so attackers don't have a way of knowing what they're attacking. Cost of executing a sandwich attack will be very costly, as they would need to publish a range of transactions in the hope that their Txs manage to sandwich an honest trade.
+
+Fee is returned to sender upon a successful Reveal. This opens up another attack vector for collators, namely censoring transactions and causing loses for users (without any cost nor benefit to the block producer).
+
+In the case of a "forgotten" Commit, the attached fee is either burnt or returned to the treasury.
+ 
+
+### Option 1: Reveal Txs Receipts / Broadcast proof
+We would like to force collators to issue a receipt of each Reveal Tx received. It's kind of like a collator committing to including a particular Tx.
+
+The problem is that as soon as the collator sees a Reveal Tx, they know the contents of the original Commit and can decide to censor the Reveal, leading to sender's punishment.
+
+Maybe instead of collators submitting receipt proofs, the sender could create a "broadcast proof" and rely on at least one collator-listener being honest.
+
+If a malicious collator then omits a Reveal (and by design, they are supposed to prioritise Reveals), then another peer can "report" on them, thus causing a slash for dishonest behaviour.
```

Comment:
This is impossible.  You cannot know why a collator drops a transaction, not without some state channel.